File: Sum_Root_to_Leaf_Numbers.java
/*129. Sum Root to Leaf Numbers  QuestionEditorial Solution  My Submissions
Total Accepted: 85895
Total Submissions: 253504
Difficulty: Medium
Given a binary tree containing digits from 0-9 only, each root-to-leaf path could represent a number.

An example is the root-to-leaf path 1->2->3 which represents the number 123.

Find the total sum of all root-to-leaf numbers.

For example,

    1
   / \
  2   3
The root-to-leaf path 1->2 represents the number 12.
The root-to-leaf path 1->3 represents the number 13.

Return the sum = 12 + 13 = 25.

Subscribe to see which companies asked this question*/

/*
Solution 2:
iterative stack DFS

Time Complexity: O(n)
Space Complexity: O(logn)
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int sumNumbers(TreeNode root) {
        int sum = 0;

        Stack<TNode> stack = new Stack<>();
        stack.push(new TNode(root, root.val));

        while (!stack.isEmpty()) {
            TNode cur = stack.pop();
            if (cur.node.left == null && cur.node.right == null) {
                sum += cur.val;
            } else {
                if (cur.node.left != null) {
                    stack.push(new TNode(cur.node.left, cur.val * 10 + cur.node.left.val));
                }
                if (cur.node.right != null) {
                    stack.push(new TNode(cur.node.right, cur.val * 10 + cur.node.right.val));
                }
            }
        }

        return sum;
    }
}

// TreeNode with current path value
class TNode {
    TreeNode node;
    int val;
    TNode(TreeNode node, int val) {
        this.val = val;
        this.node = node;
    }
}

/*
Solution 1:
Once find one leaf node, current num value *10+ node.value
if this node is not a leaf node, recursively do sumTree(root, num*10+root.val) until find the leaf node
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int sumNumbers(TreeNode root) {
        return sumTree(root, 0);
    }

    public int sumTree(TreeNode root, int num){
        if(root == null) return 0;
        if(root.right == null && root.left == null){
            return num*10+root.val;
        }
        return sumTree(root.left, num*10+root.val) + sumTree(root.right, num*10+root.val);
    }
}

File: Sum_Root_to_Leaf_Numbers.swift
/*
You are given the root of a binary tree containing digits from 0 to 9 only.

Each root-to-leaf path in the tree represents a number.

For example, the root-to-leaf path 1 -> 2 -> 3 represents the number 123.
Return the total sum of all root-to-leaf numbers. Test cases are generated so that the answer will fit in a 32-bit integer.

A leaf node is a node with no children.



Example 1:


Input: root = [1,2,3]
Output: 25
Explanation:
The root-to-leaf path 1->2 represents the number 12.
The root-to-leaf path 1->3 represents the number 13.
Therefore, sum = 12 + 13 = 25.
Example 2:


Input: root = [4,9,0,5,1]
Output: 1026
Explanation:
The root-to-leaf path 4->9->5 represents the number 495.
The root-to-leaf path 4->9->1 represents the number 491.
The root-to-leaf path 4->0 represents the number 40.
Therefore, sum = 495 + 491 + 40 = 1026.


Constraints:

The number of nodes in the tree is in the range [1, 1000].
0 <= Node.val <= 9
The depth of the tree will not exceed 10.
*/

/*
Solution 2:
iterative stack DFS

Time Complexity: O(n)
Space Complexity: O(logn)
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func sumNumbers(_ root: TreeNode?) -> Int {
        guard let root = root else { return 0 }

        var sum = 0
        var stack = [(node: TreeNode, val: Int)]()
        stack.append((root, root.val))

        while !stack.isEmpty {
            let (node, val) = stack.removeLast()

            if node.left == nil && node.right == nil {
                sum += val
            } else {
                if let left = node.left {
                    stack.append((left, val*10 + left.val))
                }
                if let right = node.right {
                    stack.append((right, val*10 + right.val))
                }
            }
        }

        return sum
    }
}

/*
Solution 1:
recursive

Time Complexity: O(n)
Space Complexity: O(1)
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func sumNumbers(_ root: TreeNode?) -> Int {
        guard let root = root else { return 0 }

        var sum = 0
        check(root, root.val, &sum)
        return sum
    }

    func check(_ node: TreeNode, _ cur: Int, _ sum: inout Int) {
        if node.left == nil && node.right == nil {
            // current node is leaf node
            sum += cur
        } else {
            if let left = node.left {
                check(left, cur*10 + left.val, &sum)
            }
            if let right = node.right {
                check(right, cur*10 + right.val, &sum)
            }
        }
    }
}
